File: kotlin/src/main/java/com/arsalabangash/boltzengine/engine/expressions/ExprGenerator.kt
package com.arsalabangash.boltzengine.engine.expressions

import com.arsalabangash.boltzengine.engine.enums.Level
import com.arsalabangash.boltzengine.engine.enums.MathOperation
import com.arsalabangash.boltzengine.engine.enums.SubExprLocation
import java.util.*

class ExprGenerator {

    val randomGenerator = Random()

    /**
     * This class is generated when an [ExprToken] with a sub-expression with it's respective [bound]
     * is found, with the original expression being split to the left and right of the
     * [ExprToken] with the sub-expression
     */
    data class ExprWithSub(val bound: Int?,
                           val leftSide: ArrayList<ExprToken>,
                           val rightSide: ArrayList<ExprToken>)

    /**
     * Reduces the question of operations left in the current expression based on the location of the
     * empty tokens processed.
     */
    fun reduceOperationsBy(subExprLocation: SubExprLocation): Int {
        when (subExprLocation) {
            SubExprLocation.LEFT, SubExprLocation.RIGHT -> return 1
            SubExprLocation.BOTH -> return 2
            else -> return 0
        }
    }

    /**
     * Based on the [operationsLeft] for a particular expression, this function will return
     * the locations for the empty tokens.
     */
    fun getEmptyTokenLocations(operationsLeft: Int): SubExprLocation {
        when (operationsLeft) {
            0, 1 ->
                return SubExprLocation.NEITHER

            2 ->
                if (randomGenerator.nextDouble() > 0.50) return SubExprLocation.LEFT
                else return SubExprLocation.RIGHT

        //An equal 1/3 chance for all EmptyTokenLocations when there are more than 2 ops left
            else -> {
                val chooseNumEmpty = randomGenerator.nextDouble()
                if (chooseNumEmpty < 0.33) return SubExprLocation.BOTH
                else if (chooseNumEmpty > 0.66) return SubExprLocation.LEFT
                else return SubExprLocation.RIGHT
            }
        }
    }

    /**
     * Given ar ArrayList of MathOperations and a Level paramater indicating the difficulty, return an 
     * expression in the form of an ArrayList of ExprTokens
     *
     * The expression returned will be in reverse polish notation
     * e.g The expression (3 + (1-2)) will be [+, 3, -, 1, 2]
     */
    fun generateExpression(operations: ArrayList<MathOperation>, level: Level): ArrayList<ExprToken> {
        var operationsLeft = operations.size
        var expression = ArrayList<ExprToken>()
        var currentOp: MathOperation = operations[randomGenerator.nextInt(operations.size)]

        //Determines whether the current operation will have numbers or sub-expressions on either side
        var subExprLocation: SubExprLocation = getEmptyTokenLocations(operationsLeft)

        operationsLeft -= reduceOperationsBy(subExprLocation)
        operations.remove(currentOp)
        expression.addAll(MathExpr.createExpression(currentOp, null, level)!!
                .produceExpression(subExprLocation))
        var exprWithSub = checkSubExprTokens(expression)
        
        //While sub-expressions still exist, we'll continue to add to the expression
        while (exprWithSub != null) {
            currentOp = operations[randomGenerator.nextInt(operations.size)]
            subExprLocation = getEmptyTokenLocations(operationsLeft)
            expression = exprWithSub.leftSide

            //Add the newly generated sub-expression to the global expression
            if (exprWithSub.bound != null) {
                expression.addAll((MathExpr.createExpression(currentOp, exprWithSub.bound, level) as MathExpr)
                        .produceExpression(subExprLocation))
            } else {
                expression.addAll(MathExpr.createExpression(currentOp, null, level)!!
                        .produceExpression(subExprLocation))
            }

            //Add the right side of the previously global expression and check again for any Sub-Expression Tokens
            expression.addAll(exprWithSub.rightSide)
            operations.remove(currentOp)
            operationsLeft -= reduceOperationsBy(subExprLocation)
            exprWithSub = checkSubExprTokens(expression)
        }

        return expression
    }

    /**
     * Scans the [expressionList] for any sub-expression tokens. When one is found, the original
     * expression is converted to a [ExprWithSub]
     */
    fun checkSubExprTokens(expressionList: ArrayList<ExprToken>): ExprWithSub? {
        for (i in expressionList.indices) {
            if (expressionList[i].hasSubExpr) {
                val bound = expressionList[i].intVal
                expressionList.removeAt(i)
                val preEmptyTokens = ArrayList(expressionList.subList(0, i))
                val postEmptyTokens = ArrayList(expressionList.subList(i, expressionList.size))
                return ExprWithSub(bound, preEmptyTokens, postEmptyTokens)
            }
        }
        return null
    }
}

File: cpp/ExprGenerator.cpp
#include "ExprGenerator.hpp"
#include "EngineUtils.hpp"

using namespace std;

/* Reduces the question of operations left in the current expression based on the location
 * of the empty tokens processed
 */
int ExprGenerator::reduceOperationsBy(SubExprLocation subExprLocation) {
    switch (subExprLocation) {
        case SubExprLocation::LEFT :
            return 1;
        case SubExprLocation::RIGHT :
            return 1;
        case SubExprLocation::BOTH :
            return 2;
        default :
            return 0;
    }
}

/* Based on the [operationsLeft] for a particular expression, this function
 * will return the locations for the empty tokens.
 */
SubExprLocation ExprGenerator::getEmptyTokenLocations(int operationsLeft) {
    switch (operationsLeft) {
        case 0 :
            return SubExprLocation::NEITHER;
        case 1 :
            return SubExprLocation::NEITHER;
        case 2 :
            if (random(0, 1) > 0.50) return SubExprLocation::LEFT;
            else return SubExprLocation::RIGHT;
        default :  // An equal 1/3 chance for all EmptyTokenLocations when there are more than 2 ops left
            double chooseNumEmpty = random(0, 1);
            if (chooseNumEmpty < 0.33) return SubExprLocation::BOTH;
            else if (chooseNumEmpty > 0.66) return SubExprLocation::LEFT;
            else return SubExprLocation::RIGHT;
    }
}

/* Given an ArrayList of MathOPerations and a Level parameter indicating the difficulty,
 * returns an expression in the form of an ArrayList of ExprTokens
 *
 * The expression returned will be in reverse polish notation
 * e.g. The expression (3 + (1 - 2)) will be [+, 3, -, 1, 2]s
 */
std::vector<ExprToken> ExprGenerator::generateExpression(std::vector<MathOperation> operations,
                                                         Level level) {
    int operationsLeft = static_cast<int>(operations.size());
    vector<ExprToken> expression;
    MathOperation currentOp = operations[random_index(operations.size())];

    // Determines whether the current operation will have numbers or sub-expressions on either side
    SubExprLocation subExprLocation = getEmptyTokenLocations(operationsLeft);

    operationsLeft -= reduceOperationsBy(subExprLocation);

    vector<MathOperation>::iterator position = find(operations.begin(), operations.end(), currentOp);
    if (position != operations.end()) operations.erase(position);

    MathExpr *expr = MathExpr::ExpressionFactory::createExpression(currentOp, NULL, level);
    vector<ExprToken> newExpressions = expr->produceExpression(subExprLocation);
    expression.insert(end(expression), begin(newExpressions), end(newExpressions));

    ExprWithSub *exprWithSub = checkSubExprTokens(expression);

    // While sub-expressions still exist, we'll continue to add to the expression
    while (exprWithSub != NULL) {
        currentOp = operations[random_index(operations.size())];
        subExprLocation = getEmptyTokenLocations(operationsLeft);
        expression = exprWithSub->leftSide;

        // Add the newly generated sub expression to the global expression
        if (exprWithSub->bound != NULL)
            expr = MathExpr::ExpressionFactory::createExpression(currentOp, exprWithSub->bound, level);
        else
            expr = MathExpr::ExpressionFactory::createExpression(currentOp, NULL, level);
        newExpressions = expr->produceExpression(subExprLocation);
        expression.insert(end(expression), begin(newExpressions), end(newExpressions));

        // Add the right side of the previously global expression and
        // check again for any Sub-Expression Tokens
        expression.insert(end(expression), begin(exprWithSub->rightSide), end(exprWithSub->rightSide));
        position = find(operations.begin(), operations.end(), currentOp);
        if (position != operations.end()) operations.erase(position);
        operationsLeft -= reduceOperationsBy(subExprLocation);
        exprWithSub = checkSubExprTokens(expression);
    }

    return expression;
}

/* Scans the [expressionList] for any sub-expression tokens. W
 * When one is found, the original expression is converted to a [ExprWithSub]
 * Returns the pointer to said [ExprWithSub], and returns NULL if no sub expression is found
 */
ExprGenerator::ExprWithSub *ExprGenerator::checkSubExprTokens(std::vector<ExprToken> expressionList) {
    ExprToken *token;
    for (int i = 0; i < expressionList.size(); i++) {
        token = &expressionList.at(i);
        if (token->hasSubExpr) {
            int bound = token->intVal;
            expressionList.erase(expressionList.begin() + i);
            vector<ExprToken> preEmptyTokens;
            preEmptyTokens.insert(preEmptyTokens.begin(), expressionList.begin(), expressionList.begin() + i - 1);
            vector<ExprToken> postEmptyTokens;
            postEmptyTokens.insert(postEmptyTokens.begin(), expressionList.begin() + i, expressionList.end());

            struct ExprWithSub *expr = new struct ExprWithSub;
            expr->bound = bound;
            expr->leftSide = preEmptyTokens;
            expr->rightSide = postEmptyTokens;

            return expr;
        }
    }
    return NULL;
}
